astropy/modeling/fitting.py

```diff
@@ -1299,16 +1369,23 @@ def _convert_input(x, y, z=None, n_models=1, model_set_axis=0):
             # Obviously this is a detail of np.linalg.lstsq and should be
             # handled specifically by any fitters that use it...
             y = np.rollaxis(y, model_set_axis, y.ndim)
+            y_shape = y.shape[:-1]
         else:
             # Shape of z excluding model_set_axis
             z_shape = z.shape[:model_set_axis] + z.shape[model_set_axis + 1:]
-
-            if not (x.shape == y.shape == z_shape):
-                raise ValueError("x, y and z should have the same shape")
+    else:
+        if z is None:
+            y_shape = y.shape
+        else:
+            z_shape = z.shape
 
     if z is None:
+        if y_shape != x.shape:
+            raise ValueError("x and y should have the same shape")
         farg = (x, y)
     else:
+        if not (x.shape == y.shape == z_shape):
+            raise ValueError("x, y and z should have the same shape")
```

Comment:
This check that I added is currently blocking the case of a length-1 model set axis, which otherwise works, so we should perhaps change that and add a test for it. This happens because the criterion for dealing with the model set axis in `_convert_inputs` is `n_models > 1` rather than the actual array dimensionality.

Reply:
In the existing code, however, providing y/z/weights with a length-1 model set axis **only works for 2D models**, not 1D ones, so this usage might not actually have been intended to work, @nden? But people could still be using it and it would seem more convenient if it does work (in both cases).

Reply:
How is it useful to have length-1 weights ?

Reply:
Not `weights` specifically, also `y` or `z`. People might already be using this, eg. because they have a script that fits some range of image rows that can sometimes be a single row (and therefore ends up with a dimension of 1 unless they explicitly check for that case). I think we may as well make life easier by tolerating that case, rather than breaking people's stuff.

Reply:
I have a fix for this.

Reply:
I think that takes care of my concerns, except that it would be good to combine the loops I mentioned (but I doubt I'd be able to look at that now before the week after next).